xpc_misc.c

```diff
@@ -114,40 +113,457 @@ xpc_array_destroy(struct xpc_object *dict)
+		size_t val_size_le = 0;
+		// `- 1` because part of the size is encoded into the type byte
+		size_t val_size_size = xpc_unpack_decode_int(&val_size_le, in + *offset) - 1;
+		size_t val_size = sizeof(size_t) == 4 ? OSSwapLittleToHostInt32(val_size_le) : OSSwapLittleToHostInt64(val_size_le);
+		*offset += val_size_size;
+		xo = xpc_unpack(in + *offset, val_size, &val_packed_size, value_type);
+		if (val_packed_size != val_size) {
+			// something's up
+			debugf("val_packed_size != val_size (%zu != %zu)", val_packed_size, val_size);
+			return NULL;
+		}
+	} else {
+		xo = xpc_unpack(in + *offset, size - *offset, &val_packed_size, 0);
+	}
+	*offset += val_packed_size;
+	return xo;
+};
+
+XPC_INLINE int xpc_pack(struct xpc_object* xo, void* buf, size_t* final_size) {
+	uint8_t* out = buf;
+	__block size_t size = 1;
+
+	if (out)
+		out[0] = xo->xo_xpc_type;
+
+	switch (xo->xo_xpc_type) {
+		case _XPC_TYPE_DICTIONARY: {
+			__block size_t offset = size;
+			xpc_dictionary_apply(xo, ^bool(const char* key, xpc_object_t value) {
+				struct xpc_object* val_xo = value;
+				size_t key_size = strlen(key) + 1;
+				size += key_size;
+				if (out) {
+					strncpy(out + offset, key, key_size - 1);
+					offset += key_size;
+					out[offset - 1] = '\0';
+				}
+				xpc_pack_encode_entry(out, &size, &offset, val_xo);
+				return true;
+			});
+		} break;
+		case _XPC_TYPE_ARRAY: {
+			__block size_t offset = size;
+			xpc_array_apply(xo, ^bool(size_t idx, xpc_object_t value) {
+				struct xpc_object* val_xo = value;
+				xpc_pack_encode_entry(out, &size, &offset, val_xo);
+				return true;
+			});
+		} break;
+		case _XPC_TYPE_BOOL: {
+			if (out)
+				out[0] |= (xpc_bool_get_value(xo) ? 1 : 0) << XPC_PACK_TYPE_BITS;
+		} break;
+		case _XPC_TYPE_DATE:
+		case _XPC_TYPE_INT64: {
+			int64_t val = xo->xo_xpc_type == _XPC_TYPE_DATE ? xpc_date_get_value(xo) : xpc_int64_get_value(xo);
+			int64_t val_le = OSSwapHostToLittleInt64(val);
+			size_t val_bits = xpc_pack_encode_int_bits(&val_le, sizeof(int64_t), false);
+			size += XPC_PACK_INT_EXTRA_BYTES(val_bits);
+			if (out)
+				xpc_pack_encode_int(out, &val_le, sizeof(int64_t));
+		} break;
+		case _XPC_TYPE_UINT64: {
+			uint64_t val = xpc_uint64_get_value(xo);
+			uint64_t val_le = OSSwapHostToLittleInt64(val);
+			size_t val_bits = xpc_pack_encode_int_bits(&val_le, sizeof(uint64_t), false);
+			size += XPC_PACK_INT_EXTRA_BYTES(val_bits);
+			if (out)
+				xpc_pack_encode_int(out, &val_le, sizeof(uint64_t));
+		} break;
+		case _XPC_TYPE_DATA: {
+			size_t len = xpc_data_get_length(xo);
+			size += len;
+			if (out)
+				memcpy(out + 1, xpc_data_get_bytes_ptr(xo), len);
+		} break;
+		case _XPC_TYPE_STRING: {
+			size_t len = xpc_string_get_length(xo);
+			size += len + 1;
+			if (out) {
+				strncpy(out + 1, xpc_string_get_string_ptr(xo), len);
+				out[1 + len] = '\0';
+			}
+		} break;
+		// currently unimplemented
+		//case _XPC_TYPE_UUID: {} break;
+		case _XPC_TYPE_FD: {
```

Comment:
Is this actually used by anything you will need at the moment?

Because it can't work the way it is. I suppose it will require the not-yet-implemented fileport_makefd / fileport_makeport syscalls.

Reply:
> I suppose it will require the not-yet-implemented fileport_makefd / fileport_makeport syscalls.

I bet it uses those. liblaunch does too, so we'd need them anyway.

Why oh why didn't they also add `portfile`... that would make my life so much easier.

Reply:
> Is this actually used by anything you will need at the moment?

Not at the moment, no. I just thought it seemed simple enough to implement so I did.

The de/serialization itself doesn't actually need those calls. They're only needed by code that calls `xpc_fd_create` or `xpc_fd_dup` (i.e. code that would be consuming or generating the `fd` objects, not our de/serialization code).